src/Tester.ts

```diff
@@ -88,461 +423,412 @@ export class Tester {
-                    if (errors) {
-                        //meaning that there is a validation error
-                        if (logMode)
-                            console.log(
-                                "\x1b[36m%s\x1b[0m",
-                                "* Created request is not valid for " + actionName + "\nMessage is " + toSend + "\nError is " + errors
-                            )
-                        container.passed = false
-                        container.report.result = new Result(13, "Created message has bad format: " + JSON.stringify(errors))
-                        resolve(true)
-                    } else {
-                        if (logMode) console.log("\x1b[36m%s\x1b[0m", "* Created request is valid for: " + actionName)
+                const responseTimeStamp = new Date()
+                container.report.sent = new Payload(invokedAction[0], toSend) //sentTimeStamp, Payload
+                self.log("Invoked action " + actionName + " with data: " + JSON.stringify(toSend, null, " "))
+                if (!interaction.hasOwnProperty("output")) {
+                    // No data received as expected.
+                    if (receivedData === undefined) {
+                        self.log(actionName + " did as expected not have a return value. Test is successful.")
+                        container.report.result = new Result(201, "The event did as expected not have a return value.")
+                        return
                     }
+                    // Received data despite not expecting any.
+                    container.passed = false
+                    try {
+                        const temp: JSON = receivedData
+                    } catch (error) {
+                        // received data is not JSON
+                        self.log(actionName + "received unexpected (action did not have an output property in TD), non JSON conformal return value.")
+                        container.report.received = JSON.parse(JSON.stringify("NOT the actual data: The received data was not JSON conformal."))
+                        container.report.result = new Result(
+                            98,
+                            "Received unexpected (action did not have an output property in TD), non JSON conformal return value."
+                        )
+                        return
+                    }
+                    self.log(
+                        actionName +
+                            "received unexpected (action did not have an output property in TD) return value: " +
+                            JSON.stringify(receivedData, null, " ")
+                    )
+                    container.report.received = receivedData
+                    container.report.result = new Result(99, "Received unexpected (action did not have an output property in TD) return value.")
+                    return
```

Comment:
@egekorkan this refers to the question where you wrote you did not unerstand it. I need feedback concerning this section from `if (!interaction.hasOwnProperty("output")) {` up to here. I added an extra JSON check to make sure I do send a non JSON value...

I will ask pretty much the same question on affected part of handleReceivedData().